File: Python/stack.py
from operator import is_
from typing import TypeVar

T = TypeVar("T")

type StackNode[T] = {
    val: T,
    next: StackNode[T] | None,
}

class LinkedListStack[T]:
    def __init__(self):
        self.stack_peek: StackNode[T] | None = None
        self.stk_size: int = 0
    
    def size(self) -> int:
        return self.stk_size
    
    def is_empty(self) -> bool:
        return self.stk_size == 0

    def push(self, value: T) -> None:
        node: StackNode[T] = { val: value, next: None }
        if self.stack_peek:
            node.next = self.stack_peek
        self.stack_pee += 1
    
    def pop(self) -> T:
        if self.is_empty():
            raise Exception("Stack is empty")
        
        node = self.stack_peek
        self.stack_peek = self.stack_peek.next
        self.stk_size -= 1
        return node.val
    
    def peek(self) -> T:
        if self.is_empty():
            raise Exception("Stack is empty")
        
        return self.stack_peek.val
    
    def to_array(self) -> list[T]:
        if self.is_empty():
            return []
        
        currentNode = self.stack_peek
        nodes: list[T] = []
        while currentNode:
            nodes.append(currentNode.val)
            currentNode = currentNode.next
        return nodes

class ArrayStack[T]:
    def __init__(self):
        self.stack: list[T] = []
    
    def size(self) -> int:
        return len(self.stack)
    
    def is_empty(self) -> bool:
        return len(self.stack) == 0
    
    def push(self, value: T) -> None:
        self.stack.append(value)
    
    def pop(self) -> T:
        if self.is_empty():
            raise Exception("Stack is empty")
        
        return self.stack.pop()
    
    def peek(self) -> T:
        if self.is_empty():
            raise Exception("Stack is empty")
        
        return self.stack[-1]
    
    def to_array(self) -> list[T]:
        return self.stack.copy()

File: TS/src/stack.ts
type StackNode<T> = {
  value: T;
  next: StackNode<T> | null;
};

class LinkedListStack<T> {
  private stackPeek: StackNode<T> | null;
  private stkSize: number;

  constructor() {
    this.stackPeek = null;
    this.stkSize = 0;
  }

  get size(): number {
    return this.stkSize;
  }

  isEmpty(): boolean {
    return this.size === 0;
  }

  push(value: T): void {
    const node: StackNode<T> = { value, next: null };
    if (this.stackPeek) {
      node.next = this.stackPeek;
    }
    this.stackPeek = node;
    this.stkSize++;
  }

  pop(): T {
    const node = this.stackPeek;
    if (node) {
      this.stackPeek = this.stackPeek!.next;
      this.stkSize--;
      return node.value;
    }
    throw new Error("Stack is empty");
  }

  peek(): T {
    if (this.stackPeek) {
      return this.stackPeek.value;
    }
    throw new Error("Stack is empty");
  }

  toArray(): T[] {
    const arr: T[] = new Array<T>(this.stkSize);
    let currentNode = this.stackPeek;
    for (let i = this.stkSize - 1; i >= 0; i--) {
      arr[i] = currentNode!.value;
      currentNode = currentNode!.next;
    }
    return arr;
  }
}

class ArrayStack<T> {
  constructor(private stack: T[] = []) {}

  get size(): number {
    return this.stack.length;
  }

  isEmpty(): boolean {
    return this.stack.length === 0;
  }

  push(value: T): void {
    this.stack.push(value);
  }

  pop(): T {
    if (this.isEmpty()) throw new Error("Stack is empty");
    return this.stack.pop()!;
  }

  peek(): T {
    if (this.isEmpty()) throw new Error("Stack is empty");
    return this.stack[this.stack.length - 1];
  }

  toArray(): T[] {
    return [...this.stack];
  }
}
